controllers/database_secret_handler.go

```diff
@@ -0,0 +1,219 @@
+	if err := c.List(context.Background(), &databaseList, &client.ListOptions{Namespace: secret.Namespace}); err != nil {
+		return nil, nil, err
+	}
+	var matchedDbs []kciv1alpha1.Database
+	var matchedDbNames []string
+	for _, database := range databaseList.Items {
+		if database.Spec.SecretName == secret.Name {
+			matchedDbs = append(matchedDbs, database)
+			matchedDbNames = append(matchedDbNames, database.Name)
+		}
+	}
+	return matchedDbs, matchedDbNames, nil
+}
+
+func isFieldUpdated(dataOld map[string][]byte, dataNew map[string][]byte, keyName string) bool {
+
+	// read old and new value
+	oldValue, oldValueOk := dataOld[keyName]
+	newValue, newValueOk := dataNew[keyName]
+
+	if !oldValueOk || !newValueOk {
+		return false // values empty or do not exist
+	}
+
+	result := bytes.Compare(oldValue, newValue)
+	return result != 0 // true if values are not empty and updated
+}
+
+func isWatchedNamespace(watchNamespaces []string, ro runtime.Object) bool {
+	if watchNamespaces[0] == "" { // # it's necessary to set "" to watch cluster wide
+		return true // watch for all namespaces
+	}
+	// define object's namespace
+	objectNamespace := ""
+	database, isDatabase := ro.(*kciv1alpha1.Database)
+	if isDatabase {
+		objectNamespace = database.Namespace
+	} else {
+		secret, isSecret := ro.(*corev1.Secret)
+		if isSecret {
+			objectNamespace = secret.Namespace
+		} else {
+			logrus.Info("unknown object", "object", ro)
+			return false
+		}
+	}
+
+	// check that current namespace is watched by db-operator
+	for _, ns := range watchNamespaces {
+		if ns == objectNamespace {
+			return true
+		}
+	}
+	return false
+}
+
+func isDatabase(ro runtime.Object) bool {
+	_, isDatabase := ro.(*kciv1alpha1.Database)
+	return isDatabase
+}
+
+func isObjectUpdated(e event.UpdateEvent) bool {
+	if e.ObjectOld == nil {
+		logrus.Error(nil, "Update event has no old runtime object to update", "event", e)
+		return false
+	}
+	if e.ObjectNew == nil {
+		logrus.Error(nil, "Update event has no new runtime object for update", "event", e)
+		return false
+	}
+	// if object kind is a Database check that 'metadata.generation' field ('spec' section) has been changed
+	_, isDatabase := e.ObjectNew.(*kciv1alpha1.Database)
+	if isDatabase {
+		return e.ObjectNew.GetGeneration() != e.ObjectOld.GetGeneration()
+	}
+
+	// if object kind is a Secret check that password value has changed
+	secretNew, isSecret := e.ObjectNew.(*corev1.Secret)
+	if isSecret {
+		// if object kind is a Secret let's check that DB password inside has been updated
+		secretOld, secretOldOk := e.ObjectOld.(*corev1.Secret)
+		if !secretOldOk {
+			logrus.Error(nil, "Update Event error! ObjectOld is NOT a Secret: e.ObjectOld.GetName=", e.ObjectOld.GetName())
+			return false
+		}
+
+		engine, passwordKey, err := resolveEngine(secretNew)
+		if err != nil {
+			return false // cannot resolve DB engine, ignore Update Event
+		}
+
+		isPasswordUpdated := isFieldUpdated(secretOld.Data, secretNew.Data, passwordKey)
```

Comment:
Is it necessary to grep exactly password value? I think we can consider as it's updated only by comparing entire secret `data` field

Reply:
ah I guess it's to determine if the secret is used for database or not.